scripts/cameo-tag.py

```diff
@@ -0,0 +1,115 @@
+# SPDX-License-Identifier: GPL-2.0-or-later
+#
+# Copyright (C) 2022 Markus Stockhausen <markus.stockhausen@gmx.de>
+#
+# ./cameo-tag.py <uImageFileName> <PartitionSplitOffset>
+#
+# CAMEO tag generator used for the D-Link DGS-1210 switches. Their U-Boot
+# loader checks for the string CAMEOTAG and a checksum in the kernel and
+# rootfs partitions. If not found it complains about the boot image.
+# Nevertheless it will boot if the tags are available in the secondary
+# boot partitions. If some day we want to overwrite the original vendor 
+# partition we must have the tags in place. To solve this we insert the
+# tag two times into the kernel image.
+#
+# To understand what we do here it is helpful to explain how the original 
+# CAMEO tag generation/checking works. The firmware consists of two parts.
+# A kernel uImage (<1.5MB) and a rootfs image (<12MB) that are written to 
+# their respective mtd partitions. The default generator simply checksums
+# both parts and appends 16 bytes [<CAMEOTAG><0001><checksum>] to each part. 
+# The checksum is only an addition of all preceding bytes (b0+b1+b2+...). 
+# A tag does not interfere with any data in the images itself. During boot
+# the loader will scan all primary/secondary partitions (2*kernel, 2*rootfs) 
+# until it finds the CAMEO tag. If checksums match everything is fine. 
+# If all 4 fail we are lost. Luckily the loader does not care about where
+# the tags are located and ignores any data beyond a tag.
+#
+# The OpenWRT image consists of a kernel (>1.5MB) and a rootfs. There is
+# no chance to add CAMEO tags at the default locations. This would leave 
+# the kernel partition without a tag. So we must find suitable space.
+# 
+# Location for original kernel partition is at the end of the uImage header. 
+# We will reuse the last bytes of the IH_NAME field. This is the tricky 
+# part because we have the header CRC and the CAMEO checksum that must match
+# the whole header. uImage header CRC checksums all data expect the CRC data
+# itself. The same applies for CAMEO checksum. It checksums all data except
+# itself. Changing one of both results in a change of the other. So we will
+# brute force checksum combinations until we get an exact match between 
+# those two. This method usually converges after ~500 iterations (<0.2s) and 
+# is easier than using reverse math. After manipulation the uImage header
+# looks like this:
+#
+# [....<header_crc>....<shortened_name><00><CAMEOTAG><0001><checksum>] 
+#
+# Location for original rootfs partition is very simple. It is behind the 
+# OpenWRT compressed kernel image file that spans into the rootfs. So
+# the tag will be written somewhere to the following rootfs partition and 
+# can be found by U-Boot. The CAMEO checksum calculation must start at
+# position 0x180000 (see DLINK_KERNEL_PART_SIZE) and includes the "second"
+# half of the "split" kernel uImage. 
+
+import argparse
+import pathlib
+import socket
+import struct
+import zlib
```

Comment:
Only zlib and argparse are used, the others can be dropped

Reply:
fixed